### DISCUSSION TITLE
Inconsistent handling of moved repositories using API v4

### DISCUSSION THREAD

**Discussion by jfeltesse-mdsol** [community/community#24517] [Apps, API and Webhooks]
<p>It seems the GraphQL API returns new repository location information <em>only</em> if the former location’s user still exists.</p>
<p>For instance, using my browser, I can access old repositories and be redirected to the new location:</p>
<p><a href="https://github.com/jim/carmen" rel="nofollow noopener">https://github.com/jim/carmen</a> → <a href="https://github.com/carmen-ruby/carmen" rel="nofollow noopener">https://github.com/carmen-ruby/carmen</a></p>
<p><a href="http://github.com/mjijackson/rack-accept" rel="nofollow noopener">http://github.com/mjijackson/rack-accept</a> → <a href="https://github.com/mjackson/rack-accept" rel="nofollow noopener">https://github.com/mjackson/rack-accept</a></p>
<p>Which would indicate in both cases the moving of the repository has been done “properly” (e.g. not by creating a brand new user and git push there), allowing Github to know where to redirect to.</p>
<p>However in GraphQL the handling is different:</p>
<pre><code>query {
  repository(owner: "jim", name: "carmen") {
    nameWithOwner
  }
}
</code></pre>
<p>↓</p>
<pre><code>{
  "data": {
    "repository": {
      "nameWithOwner": "carmen-ruby/carmen"
    }
  }
}

query {
  repository(owner: "mjijackson", name: "rack-accept") {
    nameWithOwner
  }
}
</code></pre>
<p>↓</p>
<pre><code>{
  "data": {
    "repository": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "repository"
      ],
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "message": "Could not resolve to a User with the username 'mjijackson'."
    }
  ]
}
</code></pre>
<p>I assume this is a known limitation?</p>
<p>Are there any plans to make this more consistent?</p>
<p>With the current situation I have to fallback to the v3 API  just for such cases, sending more requests than it should be and having to maintain this extra logic…</p>

**Comment by lee-dohm** [comment]
<p>Thanks for reaching out.</p>
<p>I’ve passed this along to the developers who are investigating. We’ll update you when we get more information.</p>
<p>Thanks for letting us know!</p>

**Comment by jfeltesse-mdsol** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> any update on this? It’s been a year now and as far as I can tell, this is still an issue.</p>

**Comment by lee-dohm** [comment]
<p>Unfortunately, no, there have been no updates on this.</p>

**Comment by jfeltesse-mdsol** [comment]
<p>Bummer. I’ll retry in 2021…</p>

**Comment by Nosferican** [comment]
<p>You should not be using the <code>repository(owner: String!, name: String!)</code> method for any resource you plan to query subsequently. You should always get the <code>id</code> of the repository and subsequent queries use the ID through <code>node</code> or <code>nodes</code>. That way your code is robust to renaming / transfers.</p>

**Comment by jfeltesse-mdsol** [comment]
<p>I’m not sure I understand, how do you get a repo id in the first place? The only info I have at my disposal to query is the repo owner and repo name, however outdated those may be.</p>

**Comment by Nosferican** [comment]
<p>Say you have some repositories you want to track based on the slug.<br>
You have something like the following GraphQL query</p>
<pre><code class="lang-auto">query magic($owner1: String!, $name1: String!, $owner2: String!, $name2: String!) {
  _1: repository(owner: $owner1, name: $name1) {
    id
  }
  _2: repository(owner: $owner2, name: $name2) {
    id
  }
}
</code></pre>
<p>with variables</p>
<pre><code class="lang-auto">{
  "owner1": "carmen-ruby",
  "name1": "carmen",
  "owner2": "mjackson",
  "name2": "rack-accept"
}
</code></pre>
<p>which return</p>
<pre><code class="lang-auto">{
  "data": {
    "_1": {
      "id": "MDEwOlJlcG9zaXRvcnkxNTI5ODc="
    },
    "_2": {
      "id": "MDEwOlJlcG9zaXRvcnk1OTA4NTA="
    }
  }
}
</code></pre>
<p>if the repositories are ever renamed or transferred you can always get the new location through</p>
<pre><code class="lang-auto">query magic($ids: [ID!]!) {
  nodes(ids: $ids) {
    ... on Node {
      ... on Repository {
        nameWithOwner
      }
    }
  }
}
</code></pre>
<p>and query variables</p>
<pre><code class="lang-auto">{
  "ids": ["MDEwOlJlcG9zaXRvcnkxNTI5ODc=", "MDEwOlJlcG9zaXRvcnk1OTA4NTA="]
}
</code></pre>
<p>for the following payload</p>
<pre><code class="lang-auto">{
  "data": {
    "nodes": [
      {
        "nameWithOwner": "carmen-ruby/carmen"
      },
      {
        "nameWithOwner": "mjackson/rack-accept"
      }
    ]
  }
}
</code></pre>
<p>That should get you what you want moving forward. If you just need to get the “redirect” so you can adapt the code to use the global node ID you can use any HTTP library to get the redirect; no need to use the GitHub API for that. For example using <a href="https://github.com/JuliaWeb/HTTP.jl" rel="nofollow noopener">HTTP.jl</a>,</p>
<pre><code class="lang-auto">using HTTP: request
"""
    magic(slug::AbstractString)

Return current slug of the repository or `missing` if `NOT_FOUND`.
"""
function magic(slug::AbstractString)
    try
        response = request("GET", "https://github.com/$slug")
        response.request.target[2:end]
    catch err
        err.status ≠ 404 &amp;&amp; throw(err)
        missing
    end
end
</code></pre>
<pre><code class="lang-auto">magic("jim/carmen")
</code></pre>
<pre><code class="lang-auto">"carmen-ruby/carmen"
</code></pre>
<p>Let me know if that helps.</p>

**Comment by jfeltesse-mdsol** [comment]
<p>Thanks for the thorough reply!</p>
<p>The problem I face though is I have no control whatsoever of the owner/repo name pairs I’m going to get. And while the v3 API handles just fine moved repos, the issue here is v4 falls flat on its face in the one case where a repo moved <em>and</em> the former owner account doesn’t exist anymore.<br>
When used with the old owner of rack-accept, “<strong>mjijackson</strong>”, not “<strong>mjackson</strong>”, the query to get the repo IDs fails all the same:</p>
<pre><code class="lang-auto">query magic($owner1: String!, $name1: String!, $owner2: String!, $name2: String!) {
  _1: repository(owner: $owner1, name: $name1) {
    id
  }
  _2: repository(owner: $owner2, name: $name2) {
    id
  }
}
</code></pre>
<pre><code class="lang-auto">{
  "owner1": "carmen-ruby",
  "name1": "carmen",
  "owner2": "mjijackson",
  "name2": "rack-accept"
}
</code></pre>
<pre><code class="lang-auto">{
  "data": {
    "_1": {
      "id": "MDEwOlJlcG9zaXRvcnkxNTI5ODc="
    },
    "_2": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "_2"
      ],
      "locations": [
        {
          "line": 5,
          "column": 3
        }
      ],
      "message": "Could not resolve to a User with the username 'mjijackson'."
    }
  ]
}
</code></pre>
<p>This is definitely a bug affecting the GraphQL API and it’s a shame it’s been sitting on the back burner for more than a year.<br>
Granted, those cases might be rare but regardless of occurence, I still have to maintain code to access the v3 API just for that…</p>

**Comment by Nosferican** [comment]
<p>The solution is use the global node ID which is the recommended practice for all resources that one has to query subsequently. When passing the repository by owner/name it first has to do a scan to find the node ID which incurs in wasteful computing and time. I would probably first do a GET request to get the latest slug for the repo and right after get the global node (you could do it with the GitHub RESTful v3 API as well and get the node ID from that call too). Remember that the limits for the REST and GraphQL API are separate so doing a few REST API calls does not affect your GraphQL remaining points.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Thanks for your patience folks, this should now be fixed.</p>

**Comment by jfeltesse-mdsol** [comment]
<p>Do you mean fixed and released or fixed and soon-to-be-released? Because I still get the same error I described in the original post.</p>

**Comment by Nosferican** [comment]
<p>It seems it now redirects on the name so it will work for</p>
<pre><code class="lang-auto">jim/carmen -&gt; carmen-ruby/carmen
</code></pre>
<p>but still errors out if the owner changed the name</p>
<pre><code class="lang-auto">query magic($owner1: String!, $name1: String!, $owner2: String!, $name2: String!) {
  _1: repository(owner: $owner1, name: $name1) {
    id
    nameWithOwner
  }
  _2: repository(owner: $owner2, name: $name2) {
    id
    nameWithOwner
  }
}
</code></pre>
<p>with</p>
<pre><code class="lang-auto">{
  "owner1": "jim",
  "name1": "carmen",
  "owner2": "mjijackson",
  "name2": "rack-accept"
}
</code></pre>
<p>yields</p>
<pre><code class="lang-auto">{
  "data": {
    "_1": {
      "id": "MDEwOlJlcG9zaXRvcnkxNTI5ODc=",
      "nameWithOwner": "carmen-ruby/carmen"
    },
    "_2": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "_2"
      ],
      "locations": [
        {
          "line": 6,
          "column": 3
        }
      ],
      "message": "Could not resolve to a User with the username 'mjijackson'."
    }
  ]
}
</code></pre>
<p>Would be good if the “fix” could be extended to redirect usernames by login and consequently fully match the HTTP GET behavior.</p>

**Comment by AndreaGriffiths11** [comment]
<p>The fix deployed yesterday, I’ve just ran the query and you are absolutely correct. Let me talk to the team.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Thanks for your patience <a class="mention" href="/nosferican">@nosferican</a>, turns out we had to revert this change due to some other issues. The team is working on a fix to ship this once again, I’ll keep you posted here.</p>

**Comment by Nosferican** [comment]
<p>Will be working on an application for which it would be wonderful to have this functionality by then… I will probably get to it around mid August.</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hi <a class="mention" href="/nosferican">@nosferican</a> give this a go please, it should be okay now as the latest fix has been deployed :crossed_fingers:t4:</p>

**Comment by jfeltesse-mdsol** [comment]
<p>Confirmed it’s fixed so marking this as resolved. Thanks for following up <a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a> !!!</p>
